staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go

```diff
@@ -0,0 +1,454 @@
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	if w.written {
+		klog.Errorf("Hijack called after write")
+		return nil, nil, errors.New("connection has already been written to")
+	}
+	if w.hijacked {
+		klog.Errorf("Hijack called after hijack")
+		return nil, nil, errors.New("connection has already been hijacked")
+	}
+	w.hijacked = true
+	klog.V(6).Infof("Hijack returning websocket tunneling net.Conn")
+	return w.conn, nil, nil
+}
+
+// Header is delegated to the stored "http.ResponseWriter".
+func (w *tunnelingResponseWriter) Header() http.Header {
+	return w.w.Header()
+}
+
+// Write is delegated to the stored "http.ResponseWriter".
+func (w *tunnelingResponseWriter) Write(p []byte) (int, error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	if w.hijacked {
+		klog.Errorf("Write called after hijack")
+		return 0, http.ErrHijacked
+	}
+	w.written = true
+	return w.w.Write(p)
+}
+
+// WriteHeader is delegated to the stored "http.ResponseWriter".
+func (w *tunnelingResponseWriter) WriteHeader(statusCode int) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	if w.written {
+		klog.Errorf("WriteHeader called after write")
+		return
+	}
+	if w.hijacked {
+		klog.Errorf("WriteHeader called after hijack")
+		return
+	}
+	w.written = true
+
+	if statusCode == http.StatusSwitchingProtocols {
+		// 101 upgrade responses must come via the hijacked connection, not WriteHeader
+		klog.Errorf("WriteHeader called with 101 upgrade")
+		http.Error(w.w, "unexpected upgrade", http.StatusInternalServerError)
+		return
+	}
+
+	// pass through non-upgrade responses we don't need to translate
+	w.w.WriteHeader(statusCode)
+}
+
+// headerInterceptingConn wraps the tunneling "net.Conn" to drain the
+// HTTP response status/headers from the upstream SPDY connection, then use
+// that to decide how to initialize the delegate connection for writes.
+type headerInterceptingConn struct {
+	// initializableConn is delegated to for all net.Conn methods.
+	// initializableConn.Write() is not called until response headers have been read
+	// and initializableConn#InitializeWrite() has been called with the result.
+	initializableConn
+
+	lock          sync.Mutex
+	headerBuffer  []byte
+	initialized   bool
+	initializeErr error
+}
+
+// initializableConn is a connection that will be initialized before any calls to Write are made
+type initializableConn interface {
+	net.Conn
+	// InitializeWrite is called when the backend response headers have been read.
+	// backendResponse contains the parsed headers.
+	// backendResponseBytes are the raw bytes the headers were parsed from.
+	InitializeWrite(backendResponse *http.Response, backendResponseBytes []byte) error
+}
+
+const maxHeaderBytes = 1 << 20
+
+// token for normal header / body separation (\r\n\r\n, but go tolerates the leading \r being absent)
+var lfCRLF = []byte("\n\r\n")
+
+// token for header / body separation without \r (which go tolerates)
+var lfLF = []byte("\n\n")
```

Comment:
it took me more than 5 mins to find this https://github.com/golang/go/issues/53186 , did you already know about this behavior?

Reply:
nope, my unit tests flushed it out